### ISSUE TITLE
hsa_amd_signal_async_handler returning HSA_STATUS_ERROR_INVALID_SIGNAL

### ISSUE THREAD

**Issue by marcusnchow** [RadeonOpenCompute/ROCR-Runtime#123]
Hello,
I am trying to create a callback function within ROCR for a specific signal using `hsa_amd_signal_async_handler`. However I keep getting an invalid signal error, even though I am creating the signal using `hsa_amd_signal_create`. It seems like my signal is not passing the the check to see if it can be converted to an [InterruptSignal](https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/f114e4d4153f89c1e712f679b307b83ce099240d/src/core/runtime/hsa_ext_amd.cpp#L519).
I am not sure how to pass this, are there any special attributes i need to give the signal when creating to pass this check?
Thanks for the help,
Marcus

**Comment by skeelyamd** [comment]
Signals created with hsa_amd_signal_create and attribute HSA_AMD_SIGNAL_AMD_GPU_ONLY can not be used with this function as the function is a CPU agent wait and the attribute is declaring that only a GPU agent will wait on this signal.  Those created with HSA_AMD_SIGNAL_IPC ideally should be able to use this API but are not supported today because there is no efficient wait support for them in the driver.  Allowing either signal type to pass through the API will result in busy polling in the thread that responds to these events and the effective loss of a CPU core.

Does your use case correspond to one of these?

**Comment by marcusnchow** [comment]
Are you saying that both HSA_AMD_SIGNAL_IPC and HSA_AMD_SIGNAL_AMD_GPU_ONLY are not supported for `hsa_amd_signal_async_handler`?
I dont believe that i am setting the attribute to either one of those since i give attributes = 0 to signal_create and leaving it to default. 
 
At a high level i want to call a function when a kernel packet reaches the head of the queue but before it gets dispatched to the GPU and after a running kernel is completed. So what i thought of doing is dispatching a barrier packet in front of every kernel, to mark kernel launches, and register a callback function for when that barrier packet is completed. Does that make sense? I also dont know if this is the best way to go about this either?

**Comment by skeelyamd** [comment]
Yes both attributes are not supported, though is a strong case were made for IPC we may be able to raise the priority of adding that support.  Attribute 0 signals are legal to use with a handler and should be working already.  We use these in several places in the wider ROCm stack for basic support, similar to what you are doing.

To halt the GPU you would need two barrier packets as prefix to each kernel, first to trigger the callback the second to hold up the GPU while the callback runs.  Of course you might be able to optimize that in some common cases.

Given that you want to halt the GPU between packets you might be better served by using an intercept queue and actually only submitting one packet at a time to HW.  I'm assuming that you are writing some kind of tool since this would be trivial to do at the application layer.

**Comment by marcusnchow** [comment]
hm, okay so what im doing should work. Not sure why im still gett invalid_signal error then. Ive been basing my code off what i see hows its used in the runtime. I've been making my modifications in the ROCclr layer, but i dont know if that affects anything. Something like this ... 

In [VirtualGPU::submitKernel](https://github.com/ROCm-Developer-Tools/ROCclr/blob/96724e055997473b36c74a0516d1e5ad9e1ca959/device/rocm/rocvirtual.cpp#L2874)
```
  // Dispatch barrier packet into the queue
  hsa_amd_signal_create(0, 1, &gpu_device_,0, &cu_mask_barrier_packet_.completion_signal);
  dispatchBarrierPacket&cu_mask_barrier_packet_);
  hsa_status_t status =
      hsa_amd_signal_async_handler(cu_mask_barrier_packet_.completion_signal,
                                   HSA_SIGNAL_CONDITION_LT,
                                   1,
                                   (hsa_amd_signal_handler)setCUMaskCallback,
                                   reinterpret_cast<void*>(this));
  if (status != HSA_STATUS_SUCCESS) {
    printf("hsa_amd_signal_async_handler failed: 0x%lx\n",status);
  }
```
The [ dispatchBarrierPacket](https://github.com/ROCm-Developer-Tools/ROCclr/blob/96724e055997473b36c74a0516d1e5ad9e1ca959/device/rocm/rocvirtual.cpp#L914) function I am using from ROCclr. 

But it is still producing a HSA_STATUS_ERROR_INVALID_SIGNAL error. Ill keep trying, sorry if this is too much of a debugging question. 

yes, im trying to modify the runtime to set the cu mask between kernel dispatches based on number of thread blocks used and then monitor CUs that are active or available for other kernels. Im doing it in the runtime because we have a variety of workloads that we are measuring, so yeah im trying to avoid doing it in HIP.

**Comment by skeelyamd** [comment]
If you are modifying rocclr then you can probably avoid some barrier packet overhead by submitting to a deque when HW is busy and having the post packet callback pull from the deque and insert to HW in it's epilogue.  That said changing CU masks is not a speedy operation so this might not make a significant difference.

Can you confirm that the signal you are passing has a non-zero handle?

**Comment by marcusnchow** [comment]
yes the handle is non zero, and hsa_amd_signal_create returns a success
